Use modal__spinner for compose and message edit.

shuklamaneesh23 opened the pull request:
<!-- Describe your pull request here.-->

Follow up work for  PR #31782.

Fixes: #26691.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Compose UI loader:</summary>
<table>
 <tr>
<td>
Before
</td>
<td>
After
</td>
</tr>
<tr>
<td>

https://github.com/user-attachments/assets/d4d6038f-c6e1-4273-98e8-47b82816f997


</td>
<td>

https://github.com/user-attachments/assets/038b0302-4dcc-43e0-b2b0-6b6471dd8ecc


</td>
</tr>
</table>
</details>

<details>
<summary>Edit Message UI Loader:</summary>
<table>
 <tr>
<td>
Before
</td>
<td>
After
</td>
</tr>
<tr>
<td>

https://github.com/user-attachments/assets/9f68f073-f985-4cf7-bf44-11c6beb3f5f5


</td>
<td>

https://github.com/user-attachments/assets/5c83f030-b1a2-450a-83a7-f174422b77f1


</td>
</tr>
</table>
</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

shuklamaneesh23 committed:
compose: Use new spinner for send/save for compose and message edit.

Replace the loader image with a div containing the modal__spinner class
for improved and consistent loading indication.
Refactor existing callers of show_button_spinner to utilize
show_spinner.

shuklamaneesh23 committed:
loading: Rename show_spinner and hide_spinner.

Change show_spinner to show_button_spinner and hide_spinner to
hide_button_spinner.

prakhar1144 commented:
> Never use a generic term like “bug”, “fix”, or “refactor”.

See: https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-summary-part-1

shuklamaneesh23 commented:
Hello @timabbott , 
Could you please review this? 
Thanks!

shuklamaneesh23 commented:
hey @prakhar1144 , I have made the changes as per your suggestions.
Please have another look at it and let me know if any further modifications are required.
Thanks!

shuklamaneesh23 commented:
Hey @timabbott , could you please took a look at it?
This is the follow-up work you want to be done.
Thanks!

alya commented:
I'm confused -- is this PR making the UI changes shown in the PR description? If so, a "settings" prefix on the commit messages doesn't make sense, as this is not affecting the settings.

shuklamaneesh23 commented:
Hey @alya , I have updated the commit messages.
Sorry for those mistakes!

timabbott commented on web/styles/compose.css:
```diff
@@ -1740,3 +1740,9 @@ textarea.new_message_textarea {
+    display: flex;
+    align-items: center;
+    justify-content: center;
+}
```
This seems like an odd place for this to live; `modals.css` defines all the other `modal__` classes?

shuklamaneesh23 commented:
Hey @timabbott , adressed that initial review.
Please have a look at it again and let me know about other changes.
Thanks !

timabbott commented:
@sahil839 can you review this one?

evykassirer commented on web/styles/modal.css:
```diff
@@ -348,6 +348,7 @@
 
 .modal__spinner {
     display: flex;
+    align-items: center;
```
If you're changing styles that affect all spinners, do that in a separate commit and explain why you think this shouldn't cause bugs with other spinners.

evykassirer replied:
Can you confirm you've checked all places that use `modal__spinner` to ensure this doesn't break them?

evykassirer replied:
Still waiting for a response to this

shuklamaneesh23 replied:
Yes, I checked that and have added a separate commit modifying this css property.

evykassirer replied:
Can you explain how you checked enough uses of spinners to feel confident about that?

shuklamaneesh23 replied:
Should I add some of the screenshots here for the clarification?

evykassirer replied:
That would be great, thanks! Some before and after screenshots for the places you checked :)

shuklamaneesh23 replied:
In the table below, you can clearly see that before spinner was not exactly aligned in center.

| Feature                     | Before                                                                                   | After                                                                                    |
|-----------------------------|------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------|
| **Move message modal spinner** | ![Before](https://github.com/user-attachments/assets/f55e02d8-a914-4108-9e59-3af50b677c4e) | ![After](https://github.com/user-attachments/assets/21a9d600-6ba5-412e-b6db-fdd3a3eb4339) |
| **Delete attachment spinner** | ![Before](https://github.com/user-attachments/assets/5db066ef-9a2a-4855-b6b5-1e6dc8c5b401) | ![After](https://github.com/user-attachments/assets/4b3fdd34-7cbe-4115-984f-9631cb236298) |

evykassirer replied:
What about the user profile modal? It looks like that also uses `modal__spinner`. And anything that uses `dialog_widget.launch`?

shuklamaneesh23 replied:
Here is the table for the  user profile modal and some other callers of dialog_widget.launch.
There are still many more places which are exactly same as I have attached above and here.


| Feature                 | Before                                                                                   | After                                                                                    |
|-------------------------|------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------|
| **User profile modal**  | ![Before](https://github.com/user-attachments/assets/097d82fe-192a-4a6f-bda6-80cd913a462b) | ![After](https://github.com/user-attachments/assets/bd57ebb2-3b90-4eb0-a009-06e04bb123e6) |
| **Bot section**         | ![Before](https://github.com/user-attachments/assets/52132bdc-38cc-4153-853c-56830a51a64b) | ![After](https://github.com/user-attachments/assets/2989503e-ae64-4e44-afb5-c648efb83412) |
| **Alert words**         | ![Before](https://github.com/user-attachments/assets/3939dcdb-5342-42c2-9c28-b0bf99458148) | ![After](https://github.com/user-attachments/assets/cd252ce9-ca5a-4a92-af90-73395f945edd) |

shuklamaneesh23 replied:
Hi @evykassirer , I think this much after/before gifs of loading spinner is enough to conclude that previous spinner was not present exactly in the center and hence the `align: center` is not breaking anything in the codebase.
Please have a look at the other code changes.
Thanks!

shuklamaneesh23 committed:
modal: Center modal spinner.

This commit adds the align-items property to the modal__spinner class.
This property sets the spinner in the modal to the center without
changing the spinner's position.

shuklamaneesh23 commented:
Hi @evykassirer , done the suggested change.
Please have a look again.

shuklamaneesh23 commented:
Hey @evykassirer , could you please review this?

evykassirer commented:
Every commit should be runnable locally. Notice that in the first commit you change the CSS class, but since that first commit doesn't also update `show_compose_spinner`, the spinner is temporarily broken. So those commits should actually be combined! I'd also recommend merging "Remove show_button_spinner" into that same commit as well, so that you remove the function in the same context as removing the calls to it.

I'm also confused why you've used `settings_ui` as a prefix, since the spinners you're updating are only for compose and message edit.

I think once you make those changes this should be ready to go! Looks fine for me locally. I'll take another look once you fix the commit history.

Sorry for the delay in reviewing, and thanks again for working on this!

shuklamaneesh23 commented:
Hey @evykassirer , thanks for such a detailed review.
Now, I have updated the commit history in the way that you suggested.
Do let me know if I could improve it further.

evykassirer commented:
Suggestions for commit names:

(1) compose: Use new spinner for send/save for compose and message edit.
(2) loading: Rename show_spinner and hide_spinner. 
(3)  modal: Center modal spinner.

evykassirer commented:
Tim might want to followup on https://github.com/zulip/zulip/pull/32220#discussion_r1928150983 but otherwise looks good to me. Thanks!

timabbott commented:
I'm a bit concerned that the loading indicator we're consolidating on is too subtle for the send message use case. So not reviewing the implementation here until I can think more about that.

timabbott commented:
Merged the fd2b8b5e258bcda49d84d23173c61d47c03af2b0 commit.

terpimost commented:
I suggest to have either spinner of background of a button or icon animated.
https://github.com/user-attachments/assets/4bb03142-49eb-4735-be81-1392b728fb07

The actual gradient depends on what kind of button is there.
I made some clarifications in figma, but it requires to evaluate that on every button type
